### PR TITLE
Capslock handling in text box

### PR DESCRIPTION
**Pull request by plankp**
Fixes https://github.com/ppy/osu/issues/3120

Also, how would I write a test for this?

(There is a chance that this might not compile since I do not have a working compiler)

### CODE REVIEW

File: osu.Framework/Graphics/UserInterface/TextBox.cs [hunk 1]
```diff
@@ -653,6 +656,9 @@ protected override bool OnKeyDown(InputState state, KeyDownEventArgs args)
                 case Key.Enter:
                     Commit();
                     return true;
+                case Key.CapsLock:
```
Comment (ekrctb) [inline osu.Framework/Graphics/UserInterface/TextBox.cs]: This logic can't handle the cases if capslock is already enabled before the textbox is loaded / capslock is enabled during the textbox is not focused.

Comment (plankp) [reply osu.Framework/Graphics/UserInterface/TextBox.cs]: How would I check if capslock is already enabled? Is there a routine that already does this?

Comment (ekrctb) [reply osu.Framework/Graphics/UserInterface/TextBox.cs]: Input characters are passed from (a implementation of) `ITextInputSource`. What should be fixed is this, not `TextBox`.

**Comment by plankp** [comment]
I will try to do it for windows also.

**Comment by HoutarouOreki** [comment]
On windows everything works perfectly as far as I tested (spamming with/without caps lock in framework textbox tests).

**Comment by plankp** [comment]
Oh ok! What about on linux? (Not that I can do anything about it since I do not have a linux machine)

**Comment by plankp** [comment]
Also, in that case, should I make `CapslockDown` a MacOSGameWindow specific thing?

**Comment by plankp** [comment]
Ok I will close it.
@ekrctb Upon inspecting the MacOSKeyMap.cs file in the openTk repo, I realize certain keys (including capslock) is not handled. Is there a reason for this?

**Comment by ekrctb** [comment]
I know nothing about MacOS APIs and I can't test it because I don't have a Mac. So I don't know.
Please setup a working compiler and test yourself if you going to fix the issue.

**Comment by plankp** [comment]
How would I test osu framework with a custom opentk patch?

**Comment by ekrctb** [comment]
OpenTK is referenced like <https://github.com/ppy/osu-framework/blob/c1a9cd47551171619d825db2bb667d935046c3c1/osu.Framework/osu.Framework.csproj#L27>. You could change this to your local version of OpenTK. Please consult google or other people for general problems.

**Comment by swoolcock** [comment]
In macOS, modifier keys (such as caps lock) trigger a `flagsChanged` event rather than a `keyDown`/`keyUp`, which is why OpenTK does not detect it, and why there is a hack in `MacOSGameWindow`.

If you're going to "fix" it, either add the missing case to the switch in `MacOSGameWindow`, or correctly implement `flagsChanged` in OpenTK to handle all the modifier cases.  Honestly I would just do the quick fix in osu!framework, as `CocoaNativeWindow` needs a serious rewrite which will likely happen with OpenTK 4.0.

**Comment by plankp** [comment]
Does this do the job? https://github.com/ppy/opentk/pull/26

**Comment by plankp** [comment]
@swoolcock It is based on the idea that `characters` take modifier keys into account (while the old `charactersIgnoreModifier` ignores modifier keys

**Pull request closed without merging**